File: Orimath.ViewPlugins/NotifyPropertyChanged.cs
using System;
using System.ComponentModel;
using System.Runtime.CompilerServices;

namespace Orimath.ViewPlugins
{
    //public abstract class NotifyPropertyChanged : INotifyPropertyChanged
    //{
    //    public event PropertyChangedEventHandler? PropertyChanged;

    //    protected void OnPropertyChanged([CallerMemberName]string? propertyName = null)
    //    {
    //        PropertyChanged?.Invoke(this, new PropertyChangedEventArgs(propertyName));
    //    }

    //    protected bool SetValue(ref string storage, string value, [CallerMemberName] string? propertyName = null)
    //    {
    //        if (storage == value)
    //            return false;

    //        storage = value;
    //        OnPropertyChanged(propertyName);
    //        return true;
    //    }

    //    protected bool SetValue<T>(ref T storage, T value, [CallerMemberName] string? propertyName = null)
    //        where T : notnull, IEquatable<T>
    //    {
    //        if (storage.Equals(value))
    //            return false;

    //        storage = value;
    //        OnPropertyChanged(propertyName);
    //        return true;
    //    }

    //    protected bool SetValue<T>(ref T? storage, T? value, [CallerMemberName] string? propertyName = null)
    //        where T : struct
    //    {
    //        if (Nullable.Equals(storage, value))
    //            return false;

    //        storage = value;
    //        OnPropertyChanged(propertyName);
    //        return true;
    //    }

    //    protected bool SetValueRef<T>(ref T? storage, T? value, [CallerMemberName] string? propertyName = null)
    //        where T : class
    //    {
    //        if (ReferenceEquals(storage, value))
    //            return false;

    //        storage = value;
    //        OnPropertyChanged(propertyName);
    //        return true;
    //    }
    //}
}

File: Orimath.ViewPlugins/Controls/NotifyPropertyChanged.fs
namespace Orimath.Controls
open System.Collections.Generic
open System.ComponentModel
open System.Runtime.CompilerServices
open System.Runtime.InteropServices

[<AllowNullLiteral>]
type NotifyPropertyChanged() =
    let ev = Event<PropertyChangedEventHandler, PropertyChangedEventArgs>()

    [<CLIEvent>]
    member _.PropertyChanged = ev.Publish

    member this.OnPropertyChanged([<Optional; DefaultParameterValue(""); CallerMemberName>] propertyName: string) =
        ev.Trigger(this, PropertyChangedEventArgs(propertyName))

    member this.SetValue<'T>(storage: byref<'T>, value: 'T, [<Optional; DefaultParameterValue(""); CallerMemberName>] propertyName: string) =
        if EqualityComparer<'T>.Default.Equals(storage, value) then
            false
        else
            storage <- value
            this.OnPropertyChanged(propertyName)
            true

    interface INotifyPropertyChanged with
        [<CLIEvent>]
        member _.PropertyChanged = ev.Publish
